### ISSUE TITLE
Wrong implementation of RMSprop with momentum and the fix

### ISSUE THREAD

**Issue by chingyi071** [pytorch/pytorch#32545]
## 🐛 Bug

I found the implementation of RMSprop with momentum is not consistent with the one written in the Google documentation, which is used in most of their research papers.

## To Reproduce

```python
square_avg.mul_(alpha).addcmul_(1 - alpha, grad, grad)
avg = square_avg.sqrt().add_(group['eps'])
if group['momentum'] > 0:
    buf = state['momentum_buffer']
    buf.mul_(group['momentum']).addcdiv_(grad, avg)
    p.data.add_(-group['lr'], buf)
else:
    p.data.addcdiv_(-group['lr'], grad, avg)
```
The code is equivalent to

```
mean_square_t = rho * mean_square_(t-1) + (1-rho) * gradient * gradient
mom_t = momentum * mom_(t-1) + g / sqrt(mean_square_t)
variable_t = variable_(t-1) - learning_rate * mom_t
```

## Expected behavior

The implementation should put learning rate inside the scaled gradients:
```
mean_square_t = rho * mean_square_(t-1) + (1-rho) * gradient * gradient
mom_t = momentum * mom_(t-1) + learning_rate * g / sqrt(mean_square_t)
variable_t = variable_(t-1) - mom_t
```
Based on the Tensorflow's documentation: [https://www.tensorflow.org/api_docs/python/tf/keras/optimizers/RMSprop](url)

I did the following modification, which fixed the control structure but changed something.
```python
square_avg.mul_(alpha).addcmul_(1 - alpha, grad, grad)
avg = square_avg.sqrt().add_(group['eps'])
if group['momentum'] > 0:
    eff_lr = avg.reciprocal().mul(group['lr'])
    buf = state['momentum_buffer']
    buf.mul_(group['momentum']).addcmul_(eff_lr, grad)
    p.data.add_(-1, buf)

```

## Environment
 - PyTorch Version: 1.3.1

cc @vincentqb

**Comment by albanD** [comment]
Hi,

> not consistent with the one written in the paper.

Is there a paper for RMSprop? Isn't it just on some slides from a Coursera course?

> The implementation should put learning rate inside the scaled gradients

Do you have a reference that argue about this?
All the pytorch optimizers multiply the learning rate after the momentum computation (sgd in particular). As far as I know, this is just a different convention.

**Comment by chingyi071** [comment]
> Hi,
> 
> > not consistent with the one written in the paper.
> 
> Is there a paper for RMSprop? Isn't it just on some slides from a Coursera course?
> 
> > The implementation should put learning rate inside the scaled gradients
> 
> Do you have a reference that argue about this?
> All the pytorch optimizers multiply the learning rate after the momentum computation (sgd in particular). As far as I know, this is just a different convention compared to Tensorflow.

That's a typo. But Google implemented RMSprop and used them in some papers. Our RMSprop is so different and hard to reproduce those results in PyTorch.

**Comment by chingyi071** [comment]
> Hi,
> 
> > not consistent with the one written in the paper.
> 
> Is there a paper for RMSprop? Isn't it just on some slides from a Coursera course?
> 
> > The implementation should put learning rate inside the scaled gradients
> 
> Do you have a reference that argue about this?
> All the pytorch optimizers multiply the learning rate after the momentum computation (sgd in particular). As far as I know, this is just a different convention compared to Tensorflow.

But you're right. That's not a standard. I have found some papers put momentum in gradient instead of momentum buffer. For example, [https://arxiv.org/pdf/1609.04747.pdf](url)

Empirically, putting learning rate in momentum will be unstable at the beginning of the training since the initial gradients are large and noisy.

**Comment by albanD** [comment]
From a very practical point of view, I think there is no consensus that one is better than the other (as you mentionned above, some empirical experiment argue that our version is more stable). So it is not clear which one should be implemented if we created a branch new library.

That being said, changing the current behavior of the pytorch optimizers would be very confusing for a lot of users as they will have very different behavior.

If you argue that it is hard to implement the other version based on our optimizer, I would be ok to add a new option to RMSprop (and SGD) that allows you to choose between the two. With the default behavior being the current one.
cc @vincentqb do you think that would be useful?

**Comment by vincentqb** [comment]
It would be nice in general for different libraries like PyTorch and others to have similar implementations, but we also need to protect some backward compatibility. Since there is no strong consensus here, I'm not in favor of this change.

> If you argue that it is hard to implement the other version based on our optimizer, I would be ok to add a new option to RMSprop (and SGD) that allows you to choose between the two. With the default behavior being the current one.
> cc @vincentqb do you think that would be useful?

I'm not in favor of adding more toggles to the optimizers than needed. I prefer that users are encouraged to experiment and modify the optimizers themselves.

Thanks for the input! Based on the discussion above, I will close this issue. Please feel free to re-open if needed.

**Comment by huangbiubiu** [comment]
@vincentqb When I train MobileNet v2 (implemented by torchvision) with RMSProp, the network does not converge at all. Switch to SGD fix the problem. https://github.com/marvis/pytorch-mobilenet reports a similar result on MobileNet v1 (rmsprop: top1 0.104 top5 0.494). 

In [MobileNet v2 paper](https://arxiv.org/abs/1801.04381), they use RMSProp to train the network on TensorFlow and get a good result. It's acceptable if RMSProp got a slightly worse result than other optimizers, but totally does not work is surprising.

Does this mean there is something wrong with the implementation of RMSProp in PyTorch?

**Comment by vincentqb** [comment]
The convergence of RMSProp depends on its parameter, and the exact implementation. RMSProp has never been published in an fixed form, so there are a few possible implementations with slightly different convergence and parameters.

That being said could you show us a minimal example where pytorch's implementation of RMSProp does not converge, and say this [implementation](https://gist.github.com/vincentqb/28f95d2507b7098a3ab6d07ed5d0f88f) does? The latter should be closer to tensorflow's and what is discussed above.

**Comment by huangbiubiu** [comment]
> The convergence of RMSProp depends on its parameter, and the exact implementation. RMSProp has never been published in an fixed form, so there are a few possible implementations with slightly different convergence and parameters.
> 
> That being said could you show us a minimal example where pytorch's implementation of RMSProp does not converge, and say this [implementation](https://gist.github.com/vincentqb/28f95d2507b7098a3ab6d07ed5d0f88f) does? The latter should be closer to tensorflow's and what is discussed above.

Thanks for your code! I conduct experiments (with PyTorch [official example on ImageNet](https://github.com/pytorch/examples/tree/master/imagenet)  ) on SGD, PyTorch RMSprop, and your RMSprop. SGD works well, but both RMSprop does not work at all. I tried to use different hyper-parameters (smaller learning rate, larger batch size, ...), but the loss still **not decrease at all** with both RMSprop. 

I uploaded [training logs](https://drive.google.com/drive/folders/1jzBKqX82g1Gw5-i1syvVIvd8iEeBi-9P?usp=sharing) and [code](https://github.com/huangbiubiu/mobilenet_pytorch_rmsprop). I want to know what's going on here, could you give me some suggestions?

**Comment by vincentqb** [comment]
Have you tried the modification mentioned above?

> ```python
> square_avg.mul_(alpha).addcmul_(1 - alpha, grad, grad)
> avg = square_avg.sqrt().add_(group['eps'])
> if group['momentum'] > 0:
>     eff_lr = avg.reciprocal().mul(group['lr'])
>     buf = state['momentum_buffer']
>     buf.mul_(group['momentum']).addcmul_(eff_lr, grad)
>     p.data.add_(-1, buf)
> ```

**Comment by huangbiubiu** [comment]
> Have you tried the modification mentioned above?
> 
> > ```python
> > square_avg.mul_(alpha).addcmul_(1 - alpha, grad, grad)
> > avg = square_avg.sqrt().add_(group['eps'])
> > if group['momentum'] > 0:
> >     eff_lr = avg.reciprocal().mul(group['lr'])
> >     buf = state['momentum_buffer']
> >     buf.mul_(group['momentum']).addcmul_(eff_lr, grad)
> >     p.data.add_(-1, buf)
> > ```

Yes. I copied your code in your [gist](https://gist.github.com/vincentqb/28f95d2507b7098a3ab6d07ed5d0f88f) in my code ([here](https://github.com/huangbiubiu/mobilenet_pytorch_rmsprop/blob/master/rmsprop_tf.py)). I call it `rmsprop_tf`. You can see training logs [here](https://drive.google.com/drive/folders/1jzBKqX82g1Gw5-i1syvVIvd8iEeBi-9P).

**Comment by vmmm123** [comment]
> here

hey, I encounter the same problem. Have u fix it?
